### ISSUE TITLE
Add support for multiple sizes of huge pages

### ISSUE THREAD

**Issue by JamesKingdon** [eclipse/openj9#1418]
As far as I can see we only support a legacy view of page sizes where there is a default page size (typically 4K) and a single huge page size (typically either 4M or 1G on Linux). The size of the huge pages looks to be parsed from /proc/meminfo Hugepagesize: entry. In modern systems that can have multiple pools of hugepages of different sizes, the singular value in /proc/meminfo corresponds to the default hugepage size.

The extended large page command line options
`-Xlp:objectheap:pagesize=<size> -Xlp:codecache:pagesize=<size>`

lend themselves to using multiple large page sizes. For example, on many systems it is likely that using 1G pages for the heap and 2M pages for the codecache would be an optimal choice. Even for the case of the simpler combined option `-Xlp<size>` it would be very useful to be able to choose the non-default hugepage size, which is currently not possible.

**Comment by JamesKingdon** [comment]
Tried prototyping this, and it is not as easy as advertised! So far I haven't been able to get a shmget call to return any memory from the non-default pool.

**Comment by andrewcraik** [comment]
Agreed that they are for a special situation, but if you have an app server that is being tuned to maximize throughput where you are not concerned about footprint and want to carve a machine up in an even manner with minimal TLB issues they can help. At least allowing the JVM to be configured more flexibly will allow for more experimentation and for people who benefit from these settings to have more flexibility in how to set them.

**Comment by JamesKingdon** [comment]
I made a working prototype where I can assign 4k, 2m and 1g pages in the same running JVM for any of heap, code cache and gc metadata, so there seem to be no fundamental problems.

**Comment by JamesKingdon** [comment]
@mpirvu I realised that assigning 2M pages to the code cache stopped working at some point. I started debugging and ended up concluding that the suggested address returned by chooseCacheStartAddress was most likely already mapped. All 12 probe points in the 24M window returned `Invalid argument`.
Is there any particular reason to expect the addresses below the jit library to be free?

**Comment by JamesKingdon** [comment]
I found some problems with the way the original code cache allocation falls back when it can't map at the preferred address. I fixed those and things are now more sane on my tests. I haven't had a chance to look at the new version from #1436 yet, but I suspect there's room to simplify the logic and make it more robust. Possibly we are carrying some complexity from the 32 bit case that is making life unnecessarily difficult when running in 64 bit space.

**Comment by mpirvu** [comment]
At least for pxa6480sr6-20180301_01 allocating large pages (2MB) for the code cache works for me:

[root@ivybridgedocker bin]# /home/mpirvu/sdks/pxa6480sr6-20180301_01/jre/bin/java -Xlp -Xjit:verbose={codeCache} -version

#CODECACHE:  The code cache repository was allocated between addresses 00007F6843A00000 and 00007F6845200000 to be near the VM/JIT modules to avoid trampolines
#CODECACHE:  allocated code cache segment of size 268435456
#CODECACHE:  allocateCodeCacheRepository: size=268435456 heapBase=00007F6842781000 heapAlloc=00007F6842781008 heapTop=00007F6852781000
#CODECACHE:  carved size=2097144 range: 00007F6842781008-00007F6842981000

**Comment by JamesKingdon** [comment]
Yes, it works sometimes, specifically iff the first attempt succeeds at someaddress - codecache size (roughly). But because the strict pagesize flag wasn't set, if that doesn't succeed you immediately get an buffer allocated on 4k pages, and because the returned value is !null the rest of the retry code doesn't get run. You may even have 4k pages in the output above, as it doesn't specify.

**Comment by JamesKingdon** [comment]
The fix is to set the strict pagesize flag for the first and second attempts, and to clear it for the third. However, I'm wondering what the benefit of the first attempt is, why not just start at the second?

**Comment by mpirvu** [comment]
The way I remember the VM/OMR code is that if the address we indicated is not available, the VM will increment it by 2MB (the size of the large page) and try again and again within the region we indicated.
Harry's code will turn off J9PORT_VMEM_ADDRESS_HINT for the first `allocateVirtualMemorySegmentInList` call which already has J9PORT_VMEM_STRICT_ADDRESS set.
The VM will try a few times to allocate in the region we indicate and if that fails we let the VM allocate wherever possible.

**Comment by harryyu1994** [comment]
@JamesKingdon Hi James, I think the first attempt we don't try the full 2GB range to save time, and if it did not succeed then we go with the second attempt which tries the full 2GB range. There was an issue before this change where `java -version` takes really long to finish because searching in the full range for default pages took too long.
